### ISSUE TITLE
Reproducing rec results on Inspired

### ISSUE THREAD

**Issue by RinneSz** [Alex-bupt/STEP#2]
Thanks for your excellent work! I am having trouble reproducing the rec results on Inspired reported in the paper (recall@10=0.301). I followed the execution command in rec/scripts/pretrain_inspired.sh and then rec/scripts/train_rec_inspired.sh without any augmentation to the code and the hyper-parameters. However, the results I obtained was:

Epoch 0:
{'valid/recall@1': 0.05860805860805861, 'valid/recall@10': 0.1391941391941392, 'valid/recall@50': 0.3076923076923077, 'valid/ndcg@1': 0.05860805860805861, 'valid/ndcg@10': 0.09379363147330372, 'valid/ndcg@50': 0.1310702872363639, 'valid/mrr@1': 0.05860805860805861, 'valid/mrr@10': 0.07987382909753821, 'valid/mrr@50': 0.08789180573962983, 'valid/loss': 7.901898002624511, 'epoch': 0}
{'test/recall@1': 0.0546875, 'test/recall@10': 0.19140625, 'test/recall@50': 0.32421875, 'test/ndcg@1': 0.0546875, 'test/ndcg@10': 0.11464431881904602, 'test/ndcg@50': 0.14410285651683807, 'test/mrr@1': 0.0546875, 'test/mrr@10': 0.0912310853600502, 'test/mrr@50': 0.09764114767313004, 'test/loss': 7.820733904838562, 'epoch': 0}

Epoch 1:
{'valid/recall@1': 0.054945054945054944, 'valid/recall@10': 0.16483516483516483, 'valid/recall@50': 0.31868131868131866, 'valid/ndcg@1': 0.054945054945054944, 'valid/ndcg@10': 0.10527831556159498, 'valid/ndcg@50': 0.13889045855064533, 'valid/mrr@1': 0.054945054945054944, 'valid/mrr@10': 0.08683789431393801, 'valid/mrr@50': 0.0938439212002597, 'valid/loss': 8.12632474899292, 'epoch': 1}
{'test/recall@1': 0.06640625, 'test/recall@10': 0.18359375, 'test/recall@50': 0.30078125, 'test/ndcg@1': 0.06640625, 'test/ndcg@10': 0.11988788843154907, 'test/ndcg@50': 0.14600051939487457, 'test/mrr@1': 0.06640625, 'test/mrr@10': 0.100252665579319, 'test/mrr@50': 0.10597571730613708, 'test/loss': 8.086605429649353, 'epoch': 1}

Epoch 2:
{'valid/recall@1': 0.054945054945054944, 'valid/recall@10': 0.18315018315018314, 'valid/recall@50': 0.304029304029304, 'valid/ndcg@1': 0.054945054945054944, 'valid/ndcg@10': 0.1128018557370364, 'valid/ndcg@50': 0.13954507736932664, 'valid/mrr@1': 0.054945054945054944, 'valid/mrr@10': 0.09110995002718636, 'valid/mrr@50': 0.09682187230595739, 'valid/loss': 8.539449405670165, 'epoch': 2}
{'test/recall@1': 0.0625, 'test/recall@10': 0.14453125, 'test/recall@50': 0.29296875, 'test/ndcg@1': 0.0625, 'test/ndcg@10': 0.10382596403360367, 'test/ndcg@50': 0.13622601330280304, 'test/mrr@1': 0.0625, 'test/mrr@10': 0.09059245139360428, 'test/mrr@50': 0.09736111015081406, 'test/loss': 8.56313157081604, 'epoch': 2}

Epoch 3:
{'valid/recall@1': 0.054945054945054944, 'valid/recall@10': 0.14652014652014653, 'valid/recall@50': 0.32234432234432236, 'valid/ndcg@1': 0.054945054945054944, 'valid/ndcg@10': 0.09902418227422805, 'valid/ndcg@50': 0.13862183679154505, 'valid/mrr@1': 0.054945054945054944, 'valid/mrr@10': 0.0840746561686198, 'valid/mrr@50': 0.09296250255989941, 'valid/loss': 8.962614440917969, 'epoch': 3}
{'test/recall@1': 0.08203125, 'test/recall@10': 0.16796875, 'test/recall@50': 0.3046875, 'test/ndcg@1': 0.08203125, 'test/ndcg@10': 0.12137650698423386, 'test/ndcg@50': 0.1514090746641159, 'test/mrr@1': 0.08203125, 'test/mrr@10': 0.107111856341362, 'test/mrr@50': 0.11342941969633102, 'test/loss': 8.84583854675293, 'epoch': 3}

Epoch 4:
{'valid/recall@1': 0.0695970695970696, 'valid/recall@10': 0.18315018315018314, 'valid/recall@50': 0.315018315018315, 'valid/ndcg@1': 0.0695970695970696, 'valid/ndcg@10': 0.11587784316513565, 'valid/ndcg@50': 0.14429030924926312, 'valid/mrr@1': 0.0695970695970696, 'valid/mrr@10': 0.09582097539098272, 'valid/mrr@50': 0.10149516290797418, 'valid/loss': 9.348303413391113, 'epoch': 4}
{'test/recall@1': 0.05078125, 'test/recall@10': 0.1796875, 'test/recall@50': 0.32421875, 'test/ndcg@1': 0.05078125, 'test/ndcg@10': 0.11585095524787903, 'test/ndcg@50': 0.14717450737953186, 'test/mrr@1': 0.05078125, 'test/mrr@10': 0.09539155662059784, 'test/mrr@50': 0.10183200240135193, 'test/loss': 9.127616882324219, 'epoch': 4}

The test recal@10 never exceeds 0.2. Could you please advise me on how to reproduce recall@10=0.301? Thanks!

**Comment by Alex-bupt** [comment]
Thanks for your interest. Please refer to the following parameter configuration from our run log:

```text
{'seed': 64, 'output_dir': '/root/autodl-tmp/LINK-main/output/prompt_for_recommendation', 'debug': False, 'dataset': 'inspired_gen', 'shot': 1, 'use_resp': True, 'context_max_length': 200, 'prompt_max_length': 200, 'entity_max_length': 32, 'num_workers': 0, 'tokenizer': '/root/autodl-tmp/LINK-main/models/DialoGPT-small', 'text_tokenizer': '/root/autodl-tmp/LINK-main/models/roberta-base', 'model': '/root/autodl-tmp/LINK-main/models/DialoGPT-small', 'text_encoder': '/root/autodl-tmp/LINK-main/models/roberta-base', 'num_bases': 8, 'n_prefix_rec': 20, 'prompt_encoder': '/root/autodl-tmp/LINK-main/output/pre-trained-prompt-inspired/best/', 'num_train_epochs': 5, 'max_train_steps': None, 'per_device_train_batch_size': 32, 'per_device_eval_batch_size': 64, 'gradient_accumulation_steps': 1, 'learning_rate': 0.0001, 'weight_decay': 0.01, 'max_grad_norm': None, 'num_warmup_steps': 33, 'fp16': False, 'use_wandb': False, 'entity': None, 'project': None, 'name': None, 'log_all': False}
```

For reference, the corresponding log from our run is:

```text
2025-05-08 16:18:52.533 | INFO     | __main__:<module>:303 - Saved model for epoch 2 at /root/autodl-tmp/LINK-main/output/prompt_for_recommendation/epoch_2
2025-05-08 16:18:56.201 | INFO     | __main__:<module>:341 - {'valid/recall@1': 0.14652014652014653, 'valid/recall@10': 0.2673992673992674, 'valid/recall@50': 0.4065934065934066, 'valid/ndcg@1': 0.14652014652014653, 'valid/ndcg@10': 0.20061962421123797, 'valid/ndcg@50': 0.23116948054387018, 'valid/mrr@1': 0.14652014652014653, 'valid/mrr@10': 0.1800511678059896, 'valid/mrr@50': 0.18651488587096499, 'valid/loss': 8.095121574401855, 'epoch': 2}
2025-05-08 16:18:59.512 | INFO     | __main__:<module>:387 - {'test/recall@1': 0.12109375, 'test/recall@10': 0.296875, 'test/recall@50': 0.40234375, 'test/ndcg@1': 0.12109375, 'test/ndcg@10': 0.20031540095806122, 'test/ndcg@50': 0.22382165491580963, 'test/mrr@1': 0.12109375, 'test/mrr@10': 0.17062096297740936, 'test/mrr@50': 0.17579679191112518, 'test/loss': 7.880074143409729, 'epoch': 2}
```

In our run, `test/recall@10` reached `0.296875`, which is close to the value reported in the paper.

**Comment by RinneSz** [comment]
Thanks for the information. Could you please share the hyper-parameter setting of the pre-training phase? Is it the same the pretrain_inspired.sh?

**Comment by Alex-bupt** [comment]
Yes, its same.

**Comment by RinneSz** [comment]
I followed pretrain_inspired.sh and got pre-training results like this:
`{'test/recall@1': 0.23974255832662913, 'test/recall@5': 0.37489943684633953, 'test/recall@10': 0.41914722445695896, 'test/recall@20': 0.4907481898632341, 'test/recall@50': 0.5776347546259051, 'test/ndcg@1': 0.23974255832662913, 'test/ndcg@5': 0.3125056959598753, 'test/ndcg@10': 0.32680361096358357, 'test/ndcg@20': 0.345181036264016, 'test/ndcg@50': 0.3623718605440467, 'test/mrr@1': 0.23974255832662913, 'test/mrr@5': 0.2916867907547893, 'test/mrr@10': 0.29757945765819843, 'test/mrr@20': 0.3027678387642674, 'test/mrr@50': 0.3055078409797994, 'test/loss': 5.830357098579407, 'epoch': 4}
`
And the corresponding rec results like this:
`{'test/recall@1': 0.08203125, 'test/recall@5': 0.171875, 'test/recall@10': 0.20703125, 'test/recall@20': 0.24609375, 'test/recall@50': 0.31640625, 'test/ndcg@1': 0.08203125, 'test/ndcg@5': 0.12894566357135773, 'test/ndcg@10': 0.1403016597032547, 'test/ndcg@20': 0.14984048902988434, 'test/ndcg@50': 0.16397392749786377, 'test/mrr@1': 0.08203125, 'test/mrr@5': 0.11477864533662796, 'test/mrr@10': 0.11944754421710968, 'test/mrr@20': 0.12188806384801865, 'test/mrr@50': 0.12423954159021378, 'test/loss': 9.85462737083435, 'epoch': 11}`


I was thinking maybe the pre-training phase has not converged, so I also tried to pre-train for 20 epochs, and the pre-training results are:
`{'test/recall@1': 0.4432823813354787, 'test/recall@5': 0.6637168141592921, 'test/recall@10': 0.7127916331456154, 'test/recall@20': 0.7481898632341111, 'test/recall@50': 0.7876106194690266, 'test/ndcg@1': 0.4432823813354787, 'test/ndcg@5': 0.5657082985028661, 'test/ndcg@10': 0.5817471885527705, 'test/ndcg@20': 0.5907973813511916, 'test/ndcg@50': 0.5985044078872939, 'test/mrr@1': 0.4432823813354787, 'test/mrr@5': 0.5326226772268453, 'test/mrr@10': 0.5393476417154817, 'test/mrr@20': 0.5418859774895666, 'test/mrr@50': 0.5430759420724809, 'test/loss': 4.136005139350891, 'epoch': 19}
`
Looks much better than before. The best rec performance now becomes:
`{'test/recall@1': 0.1015625, 'test/recall@5': 0.23046875, 'test/recall@10': 0.28125, 'test/recall@20': 0.31640625, 'test/recall@50': 0.38671875, 'test/ndcg@1': 0.1015625, 'test/ndcg@5': 0.17124590277671814, 'test/ndcg@10': 0.1875440776348114, 'test/ndcg@20': 0.1963111013174057, 'test/ndcg@50': 0.21048462390899658, 'test/mrr@1': 0.1015625, 'test/mrr@5': 0.15149739384651184, 'test/mrr@10': 0.15814577043056488, 'test/mrr@20': 0.16048459708690643, 'test/mrr@50': 0.16285601258277893, 'test/loss': 9.284642457962036, 'epoch': 9}`
It is closer to your reported number. However, I noticed that you are using --use_resp in your command, but most baseline methods are not using this in their experiments. After I removed --use_resp, I got rec performance like this:
`{'test/recall@1': 0.046875, 'test/recall@5': 0.13671875, 'test/recall@10': 0.18359375, 'test/recall@20': 0.2265625, 'test/recall@50': 0.33203125, 'test/ndcg@1': 0.046875, 'test/ndcg@5': 0.09324571490287781, 'test/ndcg@10': 0.1080087274312973, 'test/ndcg@20': 0.1190095767378807, 'test/ndcg@50': 0.13937948644161224, 'test/mrr@1': 0.046875, 'test/mrr@5': 0.07903645932674408, 'test/mrr@10': 0.08490048348903656, 'test/mrr@20': 0.08798962086439133, 'test/mrr@50': 0.09102723002433777, 'test/loss': 8.14266288280487, 'epoch': 3}`
which is significantly lower than using --use_resp. Is this as expected? Thank you!

**Comment by Alex-bupt** [comment]
--use_resp is needed. This hyperparameter follows the previous work. I also discovered this issue when reproducing the results (the UniCRS model,  which is the foundation model for this work). Using this hyperparameter will employ the pre-trained response template, which results in a performance gap.

**Comment by RinneSz** [comment]
UniCRS first trains the dialog module, then they generate the response using the dialog module. In other words, their --use_resp is not using the ground-truth response, but the generated response. However, I checked your code and found that you seem to be using the ground-truth response. Could this be the reason for the performance gap?

**Comment by Alex-bupt** [comment]
We also followed the training pipeline of UniCRS. However, in the later stage, we did not re-train the dialogue part. Therefore, we directly reused the responses output from the previously trained dialogue model. This resulted in some confusion when the version uploaded to GitHub was viewed. The model did not directly use the gt as the training material.

**Comment by RinneSz** [comment]
Got it. Thanks for your clarification!
